server/internal/mcp/rpc_resources_read.go

```diff
@@ -0,0 +1,255 @@
+
+	ciEnv := gateway.NewCaseInsensitiveEnv()
+	if payload.environment != "" && payload.authenticated {
+		storedEnvVars, err := env.Load(ctx, payload.projectID, gateway.Slug(payload.environment))
+		if err != nil && !errors.Is(err, gateway.ErrNotFound) {
+			return nil, oops.E(oops.CodeUnexpected, err, "failed to load environment").Log(ctx, logger)
+		}
+		for k, v := range storedEnvVars {
+			ciEnv.Set(k, v)
+		}
+	}
+
+	for k, v := range payload.mcpEnvVariables {
+		ciEnv.Set(k, v)
+	}
+
+	rw := &resourceResponseWriter{
+		body: new(bytes.Buffer),
+	}
+
+	requestBodyBytes := []byte("{}")
+	requestBytes := int64(len(requestBodyBytes))
+	var outputBytes int64
+
+	mcpURL := payload.sessionID
+	err = checkToolUsageLimits(ctx, logger, toolset.OrganizationID, toolset.AccountType, billingRepository)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		// for billing purposes we still treat fetching a resource as a type of tool call right now
+		go billingTracker.TrackToolCallUsage(context.WithoutCancel(ctx), billing.ToolCallUsageEvent{
+			OrganizationID:   toolset.OrganizationID,
+			RequestBytes:     requestBytes,
+			OutputBytes:      outputBytes,
+			ToolID:           descriptor.ID,
+			ToolName:         resourceName,
+			ResourceURI:      plan.Descriptor.URI,
+			ProjectID:        payload.projectID.String(),
+			ProjectSlug:      &descriptor.ProjectSlug,
+			OrganizationSlug: &descriptor.OrganizationSlug,
+			ToolsetSlug:      &payload.toolset,
+			ToolsetID:        &toolset.ID,
+			MCPURL:           &mcpURL,
+			MCPSessionID:     &payload.sessionID,
+			ChatID:           nil,
+			Type:             plan.BillingType,
+		})
+	}()
+
+	err = toolProxy.ReadResource(ctx, rw, strings.NewReader("{}"), ciEnv, plan)
+	if err != nil {
+		return nil, oops.E(oops.CodeUnexpected, err, "failed to execute resource call").Log(ctx, logger)
+	}
+
+	// Track output bytes
+	outputBytes = int64(rw.body.Len())
+
+	mimeType := "text/plain"
+	if plan.Function != nil && plan.Function.MimeType != "" {
+		mimeType = plan.Function.MimeType
+	}
+
+	isBinary := isBinaryMimeType(mimeType)
+
+	content := resourceContent{
+		URI:      params.URI,
+		Name:     &resourceName,
+		MimeType: &mimeType,
+		Text:     "",
+		Blob:     "",
+	}
+
+	if isBinary {
+		content.Blob = base64.StdEncoding.EncodeToString(rw.body.Bytes())
+	} else {
+		content.Text = rw.body.String()
+	}
+
+	bs, err := json.Marshal(result[resourceReadResult]{
+		ID: req.ID,
+		Result: resourceReadResult{
+			Contents: []resourceContent{content},
+		},
+	})
+	if err != nil {
+		return nil, oops.E(oops.CodeUnexpected, err, "failed to serialize resources/read result").Log(ctx, logger)
+	}
+
+	return bs, nil
+}
+
+// isBinaryMimeType determines if a MIME type represents binary content that should be base64 encoded.
+// According to MCP spec, binary resources should use base64 encoding while text resources use plain text.
+func isBinaryMimeType(mimeType string) bool {
```

Comment:
We should perhaps use a regex to catch all json/ndjson/jsonl/yaml/csv/etc... types like we do in the code generator because they will cover structured syntax suffix like `application/vnd.acme+json`